src/Frame.cc

```diff
@@ -1448,3 +1457,182 @@ void Frame::SaveFile(const std::string& root_folder, const CARTA::SaveFile& save
+        out_image->setMiscInfo(image.miscInfo());
+        out_image->setImageInfo(image.imageInfo());
+        out_image->appendLog(image.logger());
+        out_image->setUnits(image.units());
+        out_image->putSlice(temp_array, start);
+
+        // Create the mask for region
+        if (image.hasPixelMask()) {
+            casacore::Array<casacore::Bool> image_mask;
+            image.getMaskSlice(image_mask, slice);
+            out_image->makeMask("mask0", true, true);
+            casacore::Lattice<casacore::Bool>& out_image_mask = out_image->pixelMask();
+            out_image_mask.putSlice(image_mask, start);
+        }
+        success = true;
+    } catch (casacore::AipsError error) {
+        message = error.getMesg();
+    }
+
+    return success;
+}
+
+bool Frame::ExportFITSImage(casacore::ImageInterface<casacore::Float>& image, fs::path output_filename, casacore::String& message) {
+    // Remove the old image file if it has a same file name
+    return casacore::ImageFITSConverter::ImageToFITS(message, image, output_filename.string(), 64, casacore::False, casacore::True, -32,
+        1.0, -1.0, casacore::True, casacore::False, casacore::True, casacore::False, casacore::False, casacore::False, casacore::String(),
+        casacore::True);
+}
+
+void Frame::ValidateChannelStokes(std::vector<int>& channels, std::vector<int>& stokes, const CARTA::SaveFile& save_file_msg) {
+    auto image = GetImage();
+    auto image_shape = image->shape();
+
+    int channels_max = _spectral_axis > -1 ? image_shape[_spectral_axis] : 1;
+    int channels_start = 0;
+    int channels_stride = 1;
+    int channels_end = channels_max - 1;
+    int channels_length = channels_max;
+    if (save_file_msg.channels().size() > 0) {
+        channels_start = std::min(std::max(save_file_msg.channels(0), 0), channels_max - 1);
+        channels_end = std::min(std::max(save_file_msg.channels(1), channels_start), channels_max - 1);
+    }
+    int stokes_max = _stokes_axis > -1 ? image_shape[_stokes_axis] : 1;
+    int stokes_start = 0;
+    int stokes_stride = 1;
+    int stokes_end = stokes_max - 1;
+    int stokes_length = stokes_max;
+    if (save_file_msg.stokes().size() > 0) {
+        stokes_start = std::min(std::max(save_file_msg.stokes(0), 0), stokes_max - 1);
+        stokes_end = std::min(std::max(save_file_msg.stokes(1), stokes_start), stokes_max - 1);
+        stokes_stride = std::round(std::min(std::max(save_file_msg.stokes(2), 1), stokes_max - stokes_start));
+    }
+    // Ouput results
+    channels.push_back(channels_start);
+    channels.push_back(channels_end);
+    channels.push_back(channels_stride);
+    stokes.push_back(stokes_start);
+    stokes.push_back(stokes_end);
+    stokes.push_back(stokes_stride);
+}
+
+casacore::Slicer Frame::GetExportImageSlicer(const CARTA::SaveFile& save_file_msg, casacore::IPosition image_shape) {
```

Comment:
I think all of these Slicer calculations could be simplified similar to Frame::GetImageSlicer by indexing into the IPositions.  Start with a start/end/stride for the entire image, then: if you have a region, modify start/end/stride [0] and [1]; if you have a _spectral_axis, modify start/end/stride [_spectral_axis] according to the channels vector; and if you have a _stokes_axis, modify start/end/stride [_stokes_axis] according to the stokes vector.  But we could refactor this later; I have to do a bit of refactoring for supporting pV images anyway.

Reply:
Got it